File: src/backends/raster/cairo-extra/cairo-extra.c
#include "cairo-extra.h"

int32_t get_pixel(cairo_surface_t* surface, int x, int y) {
    cairo_format_t format = cairo_image_surface_get_format(surface);

    unsigned char* data = cairo_image_surface_get_data(surface);
    int stride = cairo_image_surface_get_stride(surface);

    uint32_t* p = (uint32_t*)(data + stride * y);

    return p[x];
}

void set_pixel(cairo_surface_t* surface, int x, int y, uint32_t pixel) {
    cairo_format_t format = cairo_image_surface_get_format(surface);

    unsigned char* data = cairo_image_surface_get_data(surface);
    int stride = cairo_image_surface_get_stride(surface);

    uint32_t* p = (uint32_t*)(data + stride * y);

    p[x] = pixel;
}

File: src/backends/raster/cairo-extra/cairo-extra.f90
module cairo_extra
   use, intrinsic :: iso_c_binding
   implicit none
   interface

!! Warning These procedures do not check if x and y are in bounds or if the surface is initialized.

! int32_t get_pixel(const cairo_surface_t* surface, int x, int y);
      function get_pixel(surface, x, y) bind(c)
         import :: c_ptr, c_int, c_int32_t
         implicit none
         integer(c_int32_t) :: get_pixel
         type(c_ptr), value :: surface
         integer(c_int), value :: x
         integer(c_int), value :: y
      end function

! void set_pixel(const cairo_surface_t* surface, int x, int y,int32_t pixel);
      subroutine set_pixel(surface, x, y, pixel) bind(c)
         import :: c_ptr, c_int, c_int32_t
         implicit none
         type(c_ptr), value :: surface
         integer(c_int), value :: x
         integer(c_int), value :: y
         integer(c_int32_t), value :: pixel
      end subroutine

   end interface
end module cairo_extra
